### DISCUSSION TITLE
usage grammar (matrix equality)

### DISCUSSION THREAD

**Discussion by BYS543** [yalmip/YALMIP#903] [Problems using YALMIP] [answered]
Hi, Johan

If I want to express every element of two matrix is equivalent correspondingly, can I use "Con=[A==B]", I only know the equivalence of two vectors can be expressed like this. If I cannot express it like that, is it necessary to use a for loop to express it ,just like

```
R_A=row(A); %the row number of A
Con=[];
for i=1:R_A
     Con=[Con,A(i,:)==B(i,:)];
end
```

**Comment by johanlofberg** [comment, marked as answer]
Yes, A == B is rewritten as A-B==0 following MATLABs rules for subtraction, so if A-B can be formed the statement says that all elements are equal (and it thus also works for A or B scalars)

Recent versions of MATLAB have something called singleton expansions which generalizes the subtraction of a scalar to matrices of different dimensions, but YALMIP does not support that (as it is a recipe for bugs in the code)

```
>> randn(2,3)-randn(2,1)

ans =

   -1.0234   -3.9313   -1.9594
   -0.0127   -0.0545    0.8064

>> sdpvar(2,3)-sdpvar(2,1)
Error using  -  (line 186)
Matrix dimensions must agree.
```

**Reply by BYS543** [reply]
Thanks a lot!

**Reply by BYS543** [reply]
So if I want to express every element of a matrix greater than 0, is it correct to write
```
Con=[A>=0];
is it equal to?
Con=[A>=zeros(size(A,1),size(A,2))];
```

**Reply by johanlofberg** [reply]
Yes, unless if A is symmetric since then you obtain a semidefinite constraint, so to obtain elementwise constraints in that case you would have to force a non-symmetric comparison such as `A(:)>=0`

https://yalmip.github.io/tutorial/basics/

**Reply by BYS543** [reply]
Thanks!
